src/main/vct/main/stages/Transformation.scala

```diff
@@ -122,17 +123,23 @@ class Transformation
           case (key, action) => if (pass.key == key) action(result)
         }
 
-        result = pass().dispatch(result)
-
-        result.tasks.map(_.program).flatMap(program => program.check.map(program -> _)) match {
-          case Nil => // ok
-          case errors => throw TransformationCheckError(pass, errors)
+        try {
+          result = pass().dispatch(result)
+        } catch {
+          case c @ CauseWithBadEffect(effect) =>
+            logger.error(s"An error occurred in pass ${pass.key}")
+            throw c
         }
 
         onAfterPassKey.foreach {
           case (key, action) => if (pass.key == key) action(result)
         }
 
+        result.tasks.map(_.program).flatMap(program => program.check.map(program -> _)) match {
+          case Nil => // ok
+          case errors => throw TransformationCheckError(pass, errors)
+        }
```

Comment:
Generally I would expect that attempting to print a malformed ast is less diagnostically useful than analyzing the reason the ast is malformed, did you have a case where that is not true?

Reply:
Oh, I accidentally left that there, I should put it back. Not entirely sure but I think there were one or two cases where I just got a "no successor here" or "the following key does not exist" (from succmap) error, and then it was nice to look at the output and find the ??brokenref??s. Once I saw them in context it was easy to identify the code that produced them. I think there were also a few typechecking problems caused by veymont that were nice to diagnose like this.